File: react-native/src/evaluator/VariableEvaluator.ts
const TEMPLATE_PATTERN = /\{\{([^}]+)\}\}/g;

/**
 * Coerce a JS value into a boolean using FE-renderer / native-renderer
 * semantics. Returns `null` when the value can't be mapped, so callers can
 * fall back to their own default.
 *
 *   - `boolean` → identity.
 *   - `"true"` / `"false"` strings → literal.
 *   - `"1"` / `"0"` strings → true / false (BE often ships numeric flags
 *     as strings; matches Android + iOS behavior after PR #9 / PR #11).
 *   - finite `number` → `n !== 0`.
 *   - everything else (other strings, null, undefined, objects) → `null`.
 *
 * Critically, `"0"` is treated as `false` here. The JS default
 * (`Boolean("0")` → `true`) is the opposite of what the FE renderer does,
 * so without this coercion a node with `visible: "{{flag}}"` and
 * `variables: { flag: 0 }` would render on RN but not on iOS/Android.
 */
function coerceToBool(value: unknown): boolean | null {
  if (typeof value === 'boolean') return value;
  if (typeof value === 'number' && Number.isFinite(value)) return value !== 0;
  if (typeof value === 'string') {
    if (value === 'true') return true;
    if (value === 'false') return false;
    if (value === '1') return true;
    if (value === '0') return false;
  }
  return null;
}

export class VariableEvaluator {
  constructor(private variables: Record<string, unknown>) {}

  evaluateString(template: string): string {
    return template.replace(TEMPLATE_PATTERN, (_, expr: string) => {
      const value = this._evaluateExpression(expr.trim());
      return value == null ? '' : String(value);
    });
  }

  evaluateBoolean(expression: string): boolean {
    const trimmed = expression.trim();

    if (trimmed === 'true') return true;
    if (trimmed === 'false') return false;

    // Single template variable of the form {{varName}}
    const singleVarMatch = /^\{\{([^}]+)\}\}$/.exec(trimmed);
    if (singleVarMatch) {
      const value = this._getVariable(singleVarMatch[1]!.trim());
      const coerced = coerceToBool(value);
      if (coerced != null) return coerced;
      return Boolean(value);
    }

    // Check >= and <= before > and < to avoid partial matches
    if (trimmed.includes('>=')) return this._evaluateBinaryExpr(trimmed, '>=');
    if (trimmed.includes('<=')) return this._evaluateBinaryExpr(trimmed, '<=');
    if (trimmed.includes('>')) return this._evaluateBinaryExpr(trimmed, '>');
    if (trimmed.includes('<')) return this._evaluateBinaryExpr(trimmed, '<');
    if (trimmed.includes('!=')) return this._evaluateBinaryExpr(trimmed, '!=');
    if (trimmed.includes('==')) return this._evaluateBinaryExpr(trimmed, '==');

    const value = this._resolveOperand(trimmed);
    const coerced = coerceToBool(value);
    if (coerced != null) return coerced;
    const b = this._strictBool(String(value));
    return b ?? false;
  }

  private _evaluateExpression(expression: string): unknown {
    // Ternary expression: condition ? trueValue : falseValue
    const ternaryMatch = /^(.+?)\?(.+?):(.+)$/.exec(expression);
    if (ternaryMatch) {
      const condition = ternaryMatch[1]!.trim();
      const trueVal = ternaryMatch[2]!.trim().replace(/^['"]|['"]$/g, '');
      const falseVal = ternaryMatch[3]!.trim().replace(/^['"]|['"]$/g, '');
      // Evaluate the condition as-is, not wrapped in {{}}
      return this.evaluateBoolean(condition) ? trueVal : falseVal;
    }

    // Inside {{...}}: treat the expression as a variable path; return '' if missing
    return this._getVariable(expression) ?? '';
  }

  private _evaluateBinaryExpr(expression: string, operator: string): boolean {
    const idx = expression.indexOf(operator);
    if (idx === -1) return false;
    const leftRaw = expression.slice(0, idx).trim();
    const rightRaw = expression.slice(idx + operator.length).trim();

    const leftVal = this._resolveOperand(leftRaw);
    const rightVal = this._resolveOperand(rightRaw);

    if (operator === '>' || operator === '<' || operator === '>=' || operator === '<=') {
      const left = typeof leftVal === 'number' ? leftVal : parseFloat(String(leftVal));
      const right = typeof rightVal === 'number' ? rightVal : parseFloat(String(rightVal));
      if (isNaN(left) || isNaN(right)) return false;
      switch (operator) {
        case '>': return left > right;
        case '<': return left < right;
        case '>=': return left >= right;
        case '<=': return left <= right;
        default: return false;
      }
    }

    // Equality check: convert both sides to string before comparing
    const leftStr = String(leftVal ?? '');
    const rightStr = String(rightVal ?? '');
    return operator === '==' ? leftStr === rightStr : leftStr !== rightStr;
  }

  private _resolveOperand(raw: string): unknown {
    // Template variable of the form {{varName}}
    const templateMatch = /^\{\{([^}]+)\}\}$/.exec(raw);
    if (templateMatch) {
      return this._getVariable(templateMatch[1]!.trim()) ?? '';
    }
    // Quoted string literal (single or double quotes)
    if (
      (raw.startsWith('"') && raw.endsWith('"')) ||
      (raw.startsWith("'") && raw.endsWith("'"))
    ) {
      return raw.slice(1, -1);
    }
    // Numeric literal
    const num = parseFloat(raw);
    if (!isNaN(num) && String(num) === raw) return num;
    // Boolean literal
    if (raw === 'true') return true;
    if (raw === 'false') return false;
    // Bare variable name without template syntax
    const varValue = this._getVariable(raw);
    if (varValue !== undefined) return varValue;
    // Treat anything else as an unquoted string
    return raw;
  }

  private _getVariable(name: string): unknown {
    const parts = name.split('.');
    let current: unknown = this.variables[parts[0]!];
    if (current === undefined) return undefined;

    for (const part of parts.slice(1)) {
      if (current === null || typeof current !== 'object') return undefined;
      current = (current as Record<string, unknown>)[part];
    }
    return current;
  }

  private _strictBool(value: string): boolean | null {
    if (value === 'true') return true;
    if (value === 'false') return false;
    return null;
  }
}

File: flutter/lib/src/evaluator/variable_evaluator.dart
class VariableEvaluator {
  final Map<String, dynamic> variables;

  const VariableEvaluator(this.variables);

  static final _pattern = RegExp(r'\{\{([^}]+)\}\}');

  String evaluateString(String template) {
    return template.replaceAllMapped(_pattern, (match) {
      final expr = match.group(1)!.trim();
      return _evaluateExpr(expr)?.toString() ?? '';
    });
  }

  bool evaluateBoolean(String? expr) {
    if (expr == null) return true;
    if (expr == 'true') return true;
    if (expr == 'false') return false;
    final clean = expr.startsWith('{{') && expr.endsWith('}}')
        ? expr.substring(2, expr.length - 2).trim()
        : expr;
    return _evalBoolean(clean);
  }

  dynamic _evaluateExpr(String expr) {
    final ternary = _tryTernary(expr);
    if (ternary != null) return ternary;
    final comparison = _tryComparison(expr);
    if (comparison != null) return comparison;
    return _lookupVariable(expr);
  }

  bool _evalBoolean(String expr) {
    final result = _evaluateExpr(expr);
    if (result is bool) return result;
    if (result is String) return result.isNotEmpty && result != 'false';
    if (result is num) return result != 0;
    return false;
  }

  String? _tryTernary(String expr) {
    final qIdx = expr.indexOf('?');
    final cIdx = expr.lastIndexOf(':');
    if (qIdx < 0 || cIdx <= qIdx) return null;
    final condition = expr.substring(0, qIdx).trim();
    final trueVal = expr.substring(qIdx + 1, cIdx).trim().replaceAll("'", '');
    final falseVal = expr.substring(cIdx + 1).trim().replaceAll("'", '');
    return _evalBoolean(condition) ? trueVal : falseVal;
  }

  bool? _tryComparison(String expr) {
    for (final op in ['>=', '<=', '!=', '==', '>', '<']) {
      final idx = expr.indexOf(op);
      if (idx < 0) continue;
      final left = _lookupVariable(expr.substring(0, idx).trim());
      final right = _parseValue(expr.substring(idx + op.length).trim());
      final l = _toNum(left);
      final r = _toNum(right);
      if (l != null && r != null) {
        return switch (op) {
          '>=' => l >= r,
          '<=' => l <= r,
          '!=' => l != r,
          '==' => l == r,
          '>' => l > r,
          '<' => l < r,
          _ => null,
        };
      }
      return switch (op) {
        '==' => left?.toString() == right?.toString(),
        '!=' => left?.toString() != right?.toString(),
        _ => null,
      };
    }
    return null;
  }

  dynamic _lookupVariable(String name) {
    if (name.contains('.')) {
      final parts = name.split('.');
      dynamic current = variables[parts[0]];
      for (final part in parts.skip(1)) {
        if (current is Map) {
          current = current[part];
        } else {
          return null;
        }
      }
      return current;
    }
    return variables[name];
  }

  dynamic _parseValue(String s) {
    if (s.startsWith("'") && s.endsWith("'")) return s.substring(1, s.length - 1);
    return num.tryParse(s) ?? s;
  }

  num? _toNum(dynamic v) {
    if (v is num) return v;
    if (v is String) return num.tryParse(v);
    return null;
  }
}
